File: js/nova/io/File.js
var File = function () {
	var self = this;
	
	this.fp = nova_null;
	this.location = nova_null;
	
	this.File = function (location) {
		self.this(location);
	};
	
	this.File = function (fp) {
		self.this(fp);
	};
	
	this.destroy = function () {
	};
	
	this.this = function (location) {
		self = location;
		self.fp = self.fopen(location.chars.data, "r+");
	};
	
	this.this = function (fp) {
		self = fp;
	};
	
	this.delete = function () {
		self.close();
		return self.remove(self.location.chars.data) == 0;
	};
	
	this.reopen = function () {
		self.close();
		self.fp = self.fopen(self.location.chars.data, "r+");
	};
	
	this.rewind = function () {
		self.extRewind(self.fp);
	};
	
	this.clearContents = function () {
		if (self.accessor_exists()) {
			self.fp = self.fopen(self.location.chars.data, "w");
		}
	};
	
	this.create = function () {
		if (!self.accessor_exists()) {
			self.fp = self.fopen(self.location.chars.data, "w");
			if (!self.accessor_exists()) {
				;
				return false;
			}
			self.reopen();
			if (!self.accessor_exists()) {
				;
				return false;
			}
			return true;
		}
		return false;
	};
	
	this.readAllContents = function () {
		var data;
		var line;
		data = "";
		line = self.readLine();
		while (line != null) {
			if (data.count > 0) {
				data = data.concat(new String("\n"));
			}
			data = data.concat(line);
			line = self.readLine();
		}
		return data;
	};
	
	this.readLine = function () {
		var buf;
		var size;
		var line;
		var c;
		var index;
		buf = 5;
		size = buf;
		line = [];
		c = self.getc(self.fp);
		if (c == self.EOF) {
			return null;
		}
		index = 0;
		while (c != '\n' && c != self.EOF) {
			if (index >= size) {
				size = size + buf;
				line = self.realloc(line, size);
			}
			line = c;
			c = self.getc(self.fp);
		}
		if (index >= size) {
			size++;
			line = self.realloc(line, size);
		}
		line = '\0';
		line = self.realloc(line, index);
		return new String(line);
	};
	
	this.writeLine = function (line) {
		self.write(line.concat(new String("\n")));
	};
	
	this.write = function (data) {
		self.fputs(data.chars.data, self.fp);
		self.flush();
	};
	
	this.flush = function () {
		self.fflush(self.fp);
	};
	
	this.close = function () {
		if (self.accessor_exists()) {
			self.fclose(self.fp);
		}
	};
	
	this.accessor_exists = function () {
		return self.fp != 0;
	};
	
	this.mutator_exists = function () {
	};
	
	this.accessor_maxOpenFiles = function () {
		return self.extGetMaxOpenFiles();
	};
	
	this.mutator_maxOpenFiles = function (value) {
		var min;
		var max;
		min = 20;
		max = 2048;
		if (value > max || value < min) {
			Console.writeLine(new String("Invalid max number of open files: ").concat(Int.toString(value).concat(new String("\nValid values include ").concat(Short.toString(min).concat(new String("-").concat(Short.toString(max)))))));
		} else {
			self.extSetMaxOpenFiles(value);
		}
		return value;
	};
	
	this.super = function () {
	};
	
	
};

(function () {
		var exitLog;
		exitLog = new File(new String("log").concat(Long.toString(Time.accessor_currentTimeMillis())));
})();

File: java/nova/io/File.java
package nova.io;

import java.util.Optional;
import nova.exception.ExceptionData;
import nova.exception.Exception;
import nova.exception.DivideByZeroException;
import nova.io.NovaConsole;
import nova.primitive.number.Number;
import nova.primitive.number.NovaByte;
import nova.primitive.number.NovaShort;
import nova.primitive.number.NovaInt;
import nova.primitive.number.NovaLong;
import nova.primitive.number.NovaFloat;
import nova.primitive.number.NovaDouble;
import nova.primitive.Null;
import nova.primitive.number.Char;
import nova.primitive.Bool;
import nova.datastruct.list.NovaArray;
import nova.datastruct.list.IntArray;
import nova.datastruct.list.CharArray;
import nova.datastruct.list.DoubleArray;
import nova.datastruct.list.IntRange;
import nova.thread.Thread;
import nova.thread.async.Async;
import nova.gc.GC;
import nova.math.NovaMath;
import nova.NovaObject;
import nova.NovaString;
import nova.System;
import nova.Class;
import nova.time.Time;
import nova.io.FileNotFoundException;

public class File extends NovaObject
{
	private FILE fp;
	
	public NovaString location;
	
	
	
	
	public File(NovaString location)
	{
		init(location);
	}
	
	public File(FILE fp)
	{
		init(fp);
	}
	
	public boolean accessor_exists()
	{
		return fp != 0;
	}
	
	private boolean mutator_exists()
	{
	}
	
	public int accessor_maxOpenFiles()
	{
		return extGetMaxOpenFiles();
	}
	
	public int mutator_maxOpenFiles(int value)
	{
		short min;
		short max;
		min = (short)20;
		max = (short)2048;
		if (value > max || value < min)
		{
			NovaConsole.writeLine(new NovaString("Invalid max number of open files: ").concat(NovaInt.toString(value).concat(new NovaString("\nValid values include ").concat(NovaShort.toString(min).concat(new NovaString("-").concat(NovaShort.toString(max)))))));
		}
		else
		{
			extSetMaxOpenFiles(value);
		}
		return value;
	}
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	
	public void init(NovaString location)
	{
		this.location = location;
	}
	
	public void init(FILE fp)
	{
		this.fp = fp;
	}
	
	public boolean delete()
	{
		close();
		return remove(location.chars.data) == 0;
	}
	
	public void reopen()
	{
		close();
		fp = fopen(location.chars.data, new NovaString("r+"));
	}
	
	public void rewind()
	{
		extRewind(fp);
	}
	
	public void clearContents()
	{
		if (accessor_exists())
		{
			fp = fopen(location.chars.data, new NovaString("w"));
		}
	}
	
	public boolean create()
	{
		if (!accessor_exists())
		{
			fp = fopen(location.chars.data, new NovaString("w"));
			if (!accessor_exists())
			{
				throw new FileNotFoundException(this);
				return false;
			}
			reopen();
			if (!accessor_exists())
			{
				throw new FileNotFoundException(this);
				return false;
			}
			return true;
		}
		return false;
	}
	
	public NovaString readAllContents()
	{
		NovaString data;
		NovaString line;
		data = new NovaString("");
		line = readLine();
		while (line != null)
		{
			if (data.count > 0)
			{
				data = data.concat(new NovaString("\n"));
			}
			data = data.concat(line);
			line = readLine();
		}
		return data;
	}
	
	public NovaString readLine()
	{
		int buf;
		int size;
		char[] line;
		char c;
		int index;
		buf = 5;
		size = buf;
		line = new char[size];
		c = getc(fp);
		if (c == EOF)
		{
			return null;
		}
		index = 0;
		while (c != '\n' && c != EOF)
		{
			if (index >= size)
			{
				size = size + buf;
				line = realloc(line, size);
			}
			line[index++] = c;
			c = getc(fp);
		}
		if (index >= size)
		{
			size++;
			line = realloc(line, size);
		}
		line[index++] = '\0';
		line = realloc(line, index);
		return new NovaString(line);
	}
	
	public void writeLine(NovaString line)
	{
		write(line.concat(new NovaString("\n")));
	}
	
	public void write(NovaString data)
	{
		fputs(data.chars.data, fp);
		flush();
	}
	
	public void flush()
	{
		fflush(fp);
	}
	
	public void close()
	{
		if (accessor_exists())
		{
			fclose(fp);
		}
	}
	
}
